base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -216,7 +216,7 @@ pub struct BlockchainDatabase<T>
 where T: BlockchainBackend
 {
     metadata: Arc<RwLock<ChainMetadata>>,
-    db: Arc<T>,
+    db: Arc<RwLock<T>>,
```

Comment:
Every time write access is needed a write lock guard for both is needed so I think it will be possible to wrap them into one lock. While the db write lock is held, I was hoping that the chain metadata would still be readable until the update is required, but currently it still needs some work to enable that.